types/src/transaction/webauthn.rs

```diff
@@ -0,0 +1,426 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::transaction::authenticator::AnyPublicKey;
+use anyhow::{anyhow, Result};
+use aptos_crypto::hash::CryptoHash;
+use aptos_crypto::p256_ecdsa::P256Signature;
+use aptos_crypto::{p256_ecdsa, signing_message, HashValue, Signature};
+use passkey_types::crypto::sha256;
+use passkey_types::webauthn::CollectedClientData;
+use passkey_types::Bytes;
+use serde::{Deserialize, Serialize};
+
+/// Returns `VerificationData` which composes the binary concatenation of
+/// 1. [`authenticator_data_bytes`](PartialAuthenticatorAssertionResponse) and
+/// 2. SHA-256 hash of [`client_data_json`](PartialAuthenticatorAssertionResponse),
+///
+/// See <https://www.w3.org/TR/webauthn-3/#sctn-verifying-assertion>
+#[allow(dead_code)]
+fn generate_verification_data(authenticator_data_bytes: &[u8], client_data_json: &[u8]) -> Vec<u8> {
+    // Let hash be the result of computing a hash over the clientData using SHA-256.
+    let client_data_json_hash = sha256(client_data_json);
+    // Binary concatenation of authData and hash.
+    // Note: This is compatible with signatures generated by FIDO U2F
+    // authenticators. See §6.1.2 FIDO U2F Signature Format Compatibility
+    // See <https://www.w3.org/TR/webauthn-3/#sctn-fido-u2f-sig-format-compat>
+    let verification_data: Vec<u8> = authenticator_data_bytes
+        .iter()
+        .chain(client_data_json_hash.iter())
+        .copied()
+        .collect();
+
+    verification_data
+}
+
+/// This checks that the SHA3-256 of the message (i.e., a `RawTransaction`)
+/// is equal to the actual `challenge` from a `PartialAuthenticatorAssertionResponse`
+fn verify_expected_challenge_from_message_matches_actual<T: Serialize + CryptoHash>(
+    message: &T,
+    actual_challenge: &[u8],
+) -> Result<bool> {
```

Comment:
rather than return a bool here, just return the error so you don't then have to do the stuff below

Reply:
Im not sure what you mean here?

Reply:
Done